### PR TITLE
deck: init at 1.22.0

### PR DESCRIPTION
**Pull request by liyangau**
###### Description of changes

Add new package [deck](https://github.com/Kong/deck). Users can use this package to manage their Kong API gateway configurations declaratively.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes (or backporting 23.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/networking/deck/default.nix [hunk 1]
```diff
@@ -0,0 +1,40 @@
+{ buildGoModule, lib, installShellFiles, fetchFromGitHub }:
+
+buildGoModule rec {
+  pname = "deck";
+  version = "1.22.0";
+  sha = "7447a09";
+
+  src = fetchFromGitHub {
+    owner = "Kong";
+    repo = pname;
```
Comment (drupol) [inline pkgs/applications/networking/deck/default.nix:10]: Overriding `pname` will break `src`.

```suggestion
    repo = "deck";
```

**Review by drupol — CHANGES_REQUESTED** [review]

File: pkgs/applications/networking/deck/default.nix [hunk 2]
```diff
@@ -0,0 +1,40 @@
+{ buildGoModule, lib, installShellFiles, fetchFromGitHub }:
+
+buildGoModule rec {
+  pname = "deck";
+  version = "1.22.0";
+  sha = "7447a09";
```
Comment (drupol) [inline pkgs/applications/networking/deck/default.nix:6]: What is this for?

**Review by drupol — APPROVED** [review]

File: pkgs/applications/networking/deck/default.nix (hunk 2 above)
Comment (liyangau) [reply pkgs/applications/networking/deck/default.nix:6]: Is it standard to have a `pname`, `version` in `buildGoModule`?

The `sha` variable will be pass to `ldflags`.

Comment (drupol) [reply pkgs/applications/networking/deck/default.nix:6]: > Is it standard to have a `pname`, `version` in `buildGoModule`?

By looking at other derivations, I see that it is a common practice.

Comment (drupol) [reply pkgs/applications/networking/deck/default.nix:6]: > The `sha` variable will be pass to `ldflags`.

It is a bit confusing at first, you shouldn't make it a `buildGoModule` attribute in that case.

It's better to wrap it in a `let/in` clause before `buildGoModule` so we can immediately see what are the relevant attributes.

**Review by drupol — CHANGES_REQUESTED** [review]

File: pkgs/applications/networking/deck/default.nix (hunk 2 above)
Comment (liyangau) [reply pkgs/applications/networking/deck/default.nix:6]: Thanks for the advise. Let me update this part.

File: pkgs/applications/networking/deck/default.nix [hunk 3]
```diff
@@ -0,0 +1,39 @@
+{ buildGoModule, lib, installShellFiles, fetchFromGitHub }:
+let short_hash = "7447a09";
+in buildGoModule rec {
```
Comment (drupol) [inline pkgs/applications/networking/deck/default.nix:3]:
```suggestion
let 
  short_hash = "7447a09";
in buildGoModule rec {
```

**Review by drupol — CHANGES_REQUESTED** [review]
A small nitpick and it's good to go.

**Comment by drupol** [comment]
Let's squash all the commits into one commit please.

File: pkgs/applications/networking/deck/default.nix (hunk 3 above)
Comment (drupol) [reply pkgs/applications/networking/deck/default.nix]: I think you were right, the CI is complaining now.

**Review by drupol — CHANGES_REQUESTED** [review]
I was giving a last look before merging and I noticed those things.

Please, fix them and make sure your commit log message is complying with the Nix contributor guideline: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md

Basically, your commit log message should be:

`deck: init at 1.22.0`

By using the proper commit log message, the CI tools are able to trigger the proper checks upon commits.

File: pkgs/applications/networking/deck/default.nix [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+  short_hash = "7447a09";
+in buildGoModule rec {
+  pname = "deck";
+  version = "1.22.0";
+
+  src = fetchFromGitHub {
+    owner = "Kong";
+    repo = "deck";
+    rev = "v${version}";
+    sha256 = "sha256-BCx4bw+FrnH291sp52Dz+dc6cYtoLAt8fmdF6YbmgOE=";
```
[the remark this answers is not in the document]
Comment (liyangau) [reply pkgs/applications/networking/deck/default.nix]: Thanks for the explanation. I've updated the code to use `hash` instead. 

To be honest I never understand why I need to specify *sha256* in a *sha256* hash... keyword *hash* makes more sense to me...

**Review by drupol — APPROVED** [review]

**Comment by drupol** [comment]
@GrahamcOfBorg build deck

**Review by drupol — CHANGES_REQUESTED** [review]
I think you miss an entry in `all-packages.nix` so that when we do `nix build .#deck` it should work.

**Comment by drupol** [comment]
Don't forget to squash your commits :)

**Comment by liyangau** [comment]
> Don't forget to squash your commits :)

✔️

**Comment by drupol** [comment]
Thank you for your collaboration

**Pull request merged**